// spc-api/lib/spc_tables.js
"use strict";

const SpcTables = {

    // Zone Types
    spcZoneTypes: {
        "0": "Alarm",
        "1": "Entry/Exit",
        "2": "Exit Term",
        "3": "Fire",
        "4": "Fire Exit",
        "5": "Line",
        "6": "Panic",
        "7": "Hold-up",
        "8": "Tamper",
        "9": "Technical",
        "10": "Medical",
        "11": "Keyarm",
        "12": "Unknown",
        "13": "Shunt",
        "14": "X-shunt",
        "15": "Fault",
        "16": "Lock Superv",
        "17": "Seismic",
        "18": "All OK"
    },

    // Translations for Zone types
    spcZoneTypesText: {
        en: {
            "Alarm": "Alarm",
            "Entry/Exit": "Entry/Exit",
            "Exit Term": "Exit Term",
            "Fire": "Fire",
            "Fire Exit": "Fire Exit",
            "Line": "Line",
            "Panic": "Panic",
            "Hold-up": "Hold-up",
            "Tamper": "Tamper",
            "Technical": "Technical",
            "Medical": "Medical",
            "Keyarm": "Keyarm",
            "Unknown": "Unknown",
            "Shunt": "Shunt",
            "X-shunt": "X-shunt",
            "Fault": "Fault",
            "Lock Superv": "Lock Superv",
            "Seismic": "Seismic",
            "All OK": "All OK"
        }
    },

    // Zone Input States 
    spcZoneInputs: {
        "0": "Closed",
        "1": "Open",
        "2": "Short",
        "3": "Disconnected",
        "4": "PIR Masked",
        "5": "DC Substitution",
        "6": "Sensor Missing",
        "7": "Offline"
    },

    // Tanslations for Zone Input States 
    spcZoneInputsText: {
        en: {
            "Closed": "Closed",
            "Open": "Open",
            "Short": "Short",
            "Disconnected": "Disconnected",
            "PIR Masked": "PIR Masked",
            "DC Substitution": "DC Substitution",
            "Sensor Missing": "Sensor Missing",
            "Offline": "Offline"
        }
    },


    // Zone Input Status
    spcZoneStatus: {
        "0": "OK",
        "1": "Inhibit",
        "2": "Isolate",
        "3": "Soak",
        "4": "Tamper",
        "5": "Alarm",
        "6": "OK",
        "7": "Trouble"
    },

    // Translations for Zone Input Status
    spcZoneStatusText: {
        en: {
            "OK": "OK",
            "Inhibit": "Inhibit",
            "Isolate": "Isolate",
            "Soak": "Soak",
            "Tamper": "Tamper",
            "Alarm": "Alarm",
            "Trouble": "Trouble"
        }
    },


    // Zone Area Modes
    spcAreaModes: {
        "0": "Unset",
        "1": "Partset A",
        "2": "Partset B",
        "3": "Set",
        "99": "Partly set"   // Mode differs between areas
    },

    // Translations for Zone Area Modes
    spcAreaModesText: {
        en: {
            "Unset": "Unset",
            "Partset A": "Partset A",
            "Partset B": "Partset B",
            "Set": "Set",
            "Partly set": "Partly set"
        }
    },

    // Zone Output States
    spcOutputStates: {
        "0": "Off",
        "1": "On"
    },

    // Translations for Zone Output States
    spcOutputStatesText: {
        en: {
            "Off": "Off",
            "On": "On"
        }
    }
}

module.exports = SpcTables;
